src/libcephsqlite.cc

```diff
@@ -0,0 +1,927 @@
+    vfs_context_map[std::string(dbname) + ".db"] = cc;
+  }
+  vfs_context_map_mutex.unlock();
+
+  ret = sqlite3_open_v2(filename.c_str(), &db, db_open_flags, "cephsqlite3");
+  if (ret < 0) {
+    /* db creation failed */
+    return NULL;
+  }
+
+  return db;
+}
+
+extern "C"
+void ceph_sqlite3_set_db_params(
+  const char *dbname,           /* eg. "sql" instead of "sql.db" */
+  int stripe_count,
+  int obj_size
+)
+{
+  vfs_context_map_mutex.lock();
+  CephVFSContext *cc = vfs_context_map[std::string(dbname) + ".db"];
+
+  if (cc) {
+    libradosstriper::RadosStriper *rs = cc->rs;
+
+    if (rs) {
+      rs->set_object_layout_stripe_count(stripe_count);
+      rs->set_object_layout_object_size(obj_size);
+    }
+  }
+  vfs_context_map_mutex.unlock();
+}
+
+static
+int create_lock_files(librados::IoCtx *io_ctx, const std::string &db_file, bool must_create)
+{
+  int ret = io_ctx->create(get_lock_file_name(db_file, SQLITE_LOCK_SHARED), must_create);
+  if (ret < 0 && ret != -EEXIST) {
+    // std::cerr << std::hex << pthread_self() << ":" << __FUNCTION__ << "::error: while creating:" << get_lock_file_name(db_file, SQLITE_LOCK_SHARED) << std::endl;
+    return SQLITE_ERROR;
+  }
+
+  ret = io_ctx->create(get_lock_file_name(db_file, SQLITE_LOCK_RESERVED), must_create);
+  if (ret < 0 && ret != -EEXIST) {
+    // std::cerr << std::hex << pthread_self() << ":" << __FUNCTION__ << "::error: while creating:" << get_lock_file_name(db_file, SQLITE_LOCK_RESERVED) << std::endl;
+    return SQLITE_ERROR;
+  }
+
+  ret = io_ctx->create(get_lock_file_name(db_file, SQLITE_LOCK_PENDING), must_create);
+  if (ret < 0 && ret != -EEXIST) {
+    // std::cerr << std::hex << pthread_self() << ":" << __FUNCTION__ << "::error: while creating:" << get_lock_file_name(db_file, SQLITE_LOCK_PENDING) << std::endl;
+    return SQLITE_ERROR;
+  }
+
+  ret = io_ctx->create(get_lock_file_name(db_file, SQLITE_LOCK_EXCLUSIVE), must_create);
+  if (ret < 0 && ret != -EEXIST) {
+    // std::cerr << std::hex << pthread_self() << ":" << __FUNCTION__ << "::error: while creating:" << get_lock_file_name(db_file, SQLITE_LOCK_EXCLUSIVE) << std::endl;
+    return SQLITE_ERROR;
+  }
+
+  return SQLITE_OK;
+}
+
+static
+int remove_lock_files(librados::IoCtx *io_ctx, const std::string &db_file)
+{
+  io_ctx->remove(get_lock_file_name(db_file, SQLITE_LOCK_SHARED));
+  io_ctx->remove(get_lock_file_name(db_file, SQLITE_LOCK_RESERVED));
+  io_ctx->remove(get_lock_file_name(db_file, SQLITE_LOCK_PENDING));
+  io_ctx->remove(get_lock_file_name(db_file, SQLITE_LOCK_EXCLUSIVE));
+
+  return SQLITE_OK;
+}
+
+/* input is the full URI file name*/
+static
+librados::IoCtx *get_io_ctx(const char *zPath)
+{
+  // std::cerr << std::hex << pthread_self() << ":" << __FUNCTION__ << "::" << "zPath:" << zPath << std::endl;
+
+  librados::IoCtx *io_ctx = nullptr;
+
+  vfs_context_map_mutex.lock();
+  if (vfs_context_map[std::string(zPath)])
+    io_ctx = vfs_context_map[std::string(zPath)]->io_ctx;
```

Comment:
this is an anti-pattern when using STL, would be better to put something like
```c++
if (auto found = vfs_context_map.find(zPath);
    found != vfs_context.map.end()) {
  io_ctx = found->second->io_ctx;
}
```

to avoid double lookup in the map.